File: components/esp_board_manager/devices/dev_button/dev_button.c
/*
 * SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
 * SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
 *
 * See LICENSE file for details.
 */

#include <stdlib.h>
#include "esp_log.h"
#include "esp_board_device.h"
#include "esp_board_entry.h"
#include "dev_button.h"

static const char *TAG = "DEV_BUTTON";

int dev_button_init(void *cfg, int cfg_size, void **device_handle)
{
    if (cfg == NULL || device_handle == NULL) {
        ESP_LOGE(TAG, "Invalid parameters");
        return -1;
    }
    if (cfg_size != sizeof(dev_button_config_t)) {
        ESP_LOGE(TAG, "Invalid config size");
        return -1;
    }
    esp_err_t ret = ESP_FAIL;
    dev_button_handles_t *handle = NULL;
    const dev_button_config_t *config = (const dev_button_config_t *)cfg;
    const esp_board_entry_desc_t *entry_desc = esp_board_entry_find_subtype_desc("button", config->sub_type);
    if (entry_desc == NULL) {
        ESP_LOGE(TAG, "Failed to find sub device: %s", config->sub_type);
        return -1;
    }
    ret = entry_desc->init_func((void *)config, cfg_size, (void **)&handle);
    if (ret != ESP_OK) {
        ESP_LOGE(TAG, "Failed to initialize sub device: %s", config->sub_type);
        return -1;
    }

    ESP_LOGI(TAG, "Successfully initialized button: %s, sub_type: %s",
             config->name, config->sub_type);

    *device_handle = handle;
    return 0;
}

int dev_button_deinit(void *device_handle)
{
    if (device_handle == NULL) {
        ESP_LOGE(TAG, "Invalid parameters");
        return -1;
    }
    dev_button_config_t *cfg = NULL;
    esp_board_device_get_config_by_handle(device_handle, (void **)&cfg);
    if (cfg) {
        const esp_board_entry_desc_t *desc = esp_board_entry_find_subtype_desc("button", cfg->sub_type);
        if (desc && desc->deinit_func) {
            int ret = desc->deinit_func(device_handle);
            if (ret != 0) {
                ESP_LOGE(TAG, "Sub device '%s' deinit failed with error: %d", cfg->sub_type, ret);
            } else {
                ESP_LOGI(TAG, "Sub device '%s' deinitialized successfully", cfg->sub_type);
            }
        } else {
            ESP_LOGW(TAG, "No custom deinit function found for sub type '%s'", cfg->sub_type);
        }
    }
    return 0;
}

static int dev_button_register_callback(void *handle, const void *cfg, int cfg_size, void *event_cb, void *user_data)
{
    if (handle == NULL || cfg == NULL || event_cb == NULL) {
        ESP_LOGE(TAG, "Invalid parameters");
        return -1;
    }

    if (cfg_size != sizeof(dev_button_config_t)) {
        ESP_LOGE(TAG, "Invalid config size");
        return -1;
    }

    dev_button_handles_t *button_handle = (dev_button_handles_t *)handle;
    const dev_button_config_t *config = (const dev_button_config_t *)cfg;
    button_cb_t button_event_cb = (button_cb_t)event_cb;
    esp_err_t ret = ESP_FAIL;
    // Check if user_data is NULL and use button name as default value.
    bool use_default_user_data = false;
    if (user_data == NULL) {
        use_default_user_data = true;
        ESP_LOGI(TAG, "User data is NULL, use button name as default value");
    }
    // Register enabled events based on bitfield configuration for each button.
    for (int i = 0; i < button_handle->num_buttons; i++) {
        if (use_default_user_data) {
            if (button_handle->num_buttons > 1) {
                user_data = (void *)config->sub_cfg.adc.multi.button_labels[i];
            } else {
                user_data = (void *)config->name;
            }
        }
        if (config->events_cfg.enabled_events.press_down) {
            ret = iot_button_register_cb(button_handle->button_handles[i], BUTTON_PRESS_DOWN, NULL, button_event_cb, user_data);
            ESP_BOARD_RETURN_ON_FALSE(ret == ESP_OK, -1, TAG, "Failed to register press_down callback: %s", esp_err_to_name(ret));
        }

        if (config->events_cfg.enabled_events.press_up) {
            ret = iot_button_register_cb(button_handle->button_handles[i], BUTTON_PRESS_UP, NULL, button_event_cb, user_data);
            ESP_BOARD_RETURN_ON_FALSE(ret == ESP_OK, -1, TAG, "Failed to register press_up callback: %s", esp_err_to_name(ret));
        }

        if (config->events_cfg.enabled_events.single_click) {
            ret = iot_button_register_cb(button_handle->button_handles[i], BUTTON_SINGLE_CLICK, NULL, button_event_cb, user_data);
            ESP_BOARD_RETURN_ON_FALSE(ret == ESP_OK, -1, TAG, "Failed to register single_click callback: %s", esp_err_to_name(ret));
        }

        if (config->events_cfg.enabled_events.double_click) {
            ret = iot_button_register_cb(button_handle->button_handles[i], BUTTON_DOUBLE_CLICK, NULL, button_event_cb, user_data);
            ESP_BOARD_RETURN_ON_FALSE(ret == ESP_OK, -1, TAG, "Failed to register double_click callback: %s", esp_err_to_name(ret));
        }

        if (config->events_cfg.enabled_events.long_press_hold) {
            ret = iot_button_register_cb(button_handle->button_handles[i], BUTTON_LONG_PRESS_HOLD, NULL, button_event_cb, user_data);
            ESP_BOARD_RETURN_ON_FALSE(ret == ESP_OK, -1, TAG, "Failed to register long_press_hold callback: %s", esp_err_to_name(ret));
        }

        if (config->events_cfg.enabled_events.press_repeat) {
            ret = iot_button_register_cb(button_handle->button_handles[i], BUTTON_PRESS_REPEAT, NULL, button_event_cb, user_data);
            ESP_BOARD_RETURN_ON_FALSE(ret == ESP_OK, -1, TAG, "Failed to register press_repeat callback: %s", esp_err_to_name(ret));
        }

        if (config->events_cfg.enabled_events.press_repeat_done) {
            ret = iot_button_register_cb(button_handle->button_handles[i], BUTTON_PRESS_REPEAT_DONE, NULL, button_event_cb, user_data);
            ESP_BOARD_RETURN_ON_FALSE(ret == ESP_OK, -1, TAG, "Failed to register press_repeat_done callback: %s", esp_err_to_name(ret));
        }

        if (config->events_cfg.enabled_events.press_end) {
            ret = iot_button_register_cb(button_handle->button_handles[i], BUTTON_PRESS_END, NULL, button_event_cb, user_data);
            ESP_BOARD_RETURN_ON_FALSE(ret == ESP_OK, -1, TAG, "Failed to register press_end callback: %s", esp_err_to_name(ret));
        }

        if (config->events_cfg.enabled_events.long_press_start) {
            ret = iot_button_register_cb(button_handle->button_handles[i], BUTTON_LONG_PRESS_START, NULL, button_event_cb, user_data);
            ESP_BOARD_RETURN_ON_FALSE(ret == ESP_OK, -1, TAG, "Failed to register long_press_start callback: %s", esp_err_to_name(ret));
        }

        if (config->events_cfg.enabled_events.long_press_up) {
            ret = iot_button_register_cb(button_handle->button_handles[i], BUTTON_LONG_PRESS_UP, NULL, button_event_cb, user_data);
            ESP_BOARD_RETURN_ON_FALSE(ret == ESP_OK, -1, TAG, "Failed to register long_press_up callback: %s", esp_err_to_name(ret));
        }

        if (config->events_cfg.enabled_events.long_press_start && config->events_cfg.long_press_start.count > 0 && config->events_cfg.long_press_start.durations_ms != NULL) {
            for (int j = 0; j < config->events_cfg.long_press_start.count; ++j) {
                button_event_args_t long_press_temp_args = {
                    .long_press.press_time = config->events_cfg.long_press_start.durations_ms[j],
                };
                ret = iot_button_register_cb(button_handle->button_handles[i], BUTTON_LONG_PRESS_START, &long_press_temp_args, button_event_cb, user_data);
                ESP_BOARD_RETURN_ON_FALSE(ret == ESP_OK, -1, TAG, "Failed to register long_press_start (durations_ms: %d) callback: %s", config->events_cfg.long_press_start.durations_ms[j], esp_err_to_name(ret));
            }
        }

        if (config->events_cfg.enabled_events.long_press_up && config->events_cfg.long_press_up.count > 0 && config->events_cfg.long_press_up.durations_ms != NULL) {
            for (int j = 0; j < config->events_cfg.long_press_up.count; ++j) {
                button_event_args_t long_press_temp_args = {
                    .long_press.press_time = config->events_cfg.long_press_up.durations_ms[j],
                };
                ret = iot_button_register_cb(button_handle->button_handles[i], BUTTON_LONG_PRESS_UP, &long_press_temp_args, button_event_cb, user_data);
                ESP_BOARD_RETURN_ON_FALSE(ret == ESP_OK, -1, TAG, "Failed to register long_press_up (durations_ms: %d) callback: %s", config->events_cfg.long_press_up.durations_ms[j], esp_err_to_name(ret));
            }
        }

        if (config->events_cfg.enabled_events.multi_click && config->events_cfg.multi_click.count > 0 && config->events_cfg.multi_click.click_counts != NULL) {
            for (int j = 0; j < config->events_cfg.multi_click.count; ++j) {
                button_event_args_t multiple_click_temp_args = {
                    .multiple_clicks.clicks = config->events_cfg.multi_click.click_counts[j],
                };
                ret = iot_button_register_cb(button_handle->button_handles[i], BUTTON_MULTIPLE_CLICK, &multiple_click_temp_args, button_event_cb, user_data);
                ESP_BOARD_RETURN_ON_FALSE(ret == ESP_OK, -1, TAG, "Failed to register multi_click (click_counts: %d) callback: %s", config->events_cfg.multi_click.click_counts[j], esp_err_to_name(ret));
            }
        }
    }

    ESP_LOGI(TAG, "Successfully registered callback for %s", config->name);
    return 0;
}

DEVICE_EXTRA_FUNC_REGISTER(button, dev_button_register_callback);

File: components/esp_board_manager/devices/dev_button/dev_button.py
# SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
# SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
#
# See LICENSE file for details.

#  Button device config parser
VERSION = 'v1.0.0'

import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).parent.parent.parent))

def get_includes() -> list:
    """Return list of required include headers for  button device"""
    return [
        'dev_button.h'
    ]


def _find_named_peripheral(peripherals, prefix: str):
    for periph in peripherals:
        if isinstance(periph, dict):
            periph_name = periph.get('name', '')
            if periph_name.startswith(prefix):
                return periph_name
        elif isinstance(periph, str) and periph.startswith(prefix):
            return periph
    return None


def parse(name: str, config: dict, peripherals_dict=None) -> dict:
    """Parse  button device configuration from YAML to C structure"""

    # Parse the device name - use name directly for C naming
    c_name = name.replace('-', '_')  # Convert hyphens to underscores for C naming

    # Get the device config
    device_config = config.get('config', {})
    peripherals = config.get('peripherals', [])

    # Get button type
    button_type = config.get('sub_type')
    if button_type not in ['gpio', 'adc_single', 'adc_multi', 'custom']:
        raise ValueError(f"YAML validation error in  button device: Invalid button type '{button_type}'. Valid types: ['gpio', 'adc_single', 'adc_multi', 'custom']")

    # Parse timing configuration
    long_press_time = int(device_config.get('long_press_time', 2000))
    short_press_time = int(device_config.get('short_press_time', 100))
    button_timing_cfg = {
        'long_press_time': long_press_time,
        'short_press_time': short_press_time
    }

    # Parse event configuration (Optimized - bitfield storage)
    if 'events' in device_config:
        raise ValueError(
            "YAML validation error in  button device: Legacy 'events' is no longer supported. "
            "Use 'events_cfg' under config."
        )

    events_config = device_config.get('events_cfg', {})

    # Parse enabled events bitfield
    enabled_events = {
        'press_down': int(events_config.get('press_down', True)),
        'press_up': int(events_config.get('press_up', True)),
        'single_click': int(events_config.get('single_click', True)),
        'double_click': int(events_config.get('double_click', True)),
        'multi_click': int(events_config.get('multi_click', False)),
        'long_press_start': int(events_config.get('long_press_start', True)),
        'long_press_hold': int(events_config.get('long_press_hold', False)),
        'long_press_up': int(events_config.get('long_press_up', True)),
        'press_repeat': int(events_config.get('press_repeat', False)),
        'press_repeat_done': int(events_config.get('press_repeat_done', False)),
        'press_end': int(events_config.get('press_end', False))
    }

    # Build events configuration structure
    events_cfg = {
        'enabled_events': enabled_events,
    }

    extra_configs = []
    if enabled_events['multi_click'] == 1:
        # Handle multi_click array
        multi_click_value = events_config.get('click_counts')
        if multi_click_value is not None and isinstance(multi_click_value, list):
            multi_click_counts = []
            multi_click_counts = [int(x) for x in multi_click_value]
            name = c_name + '_click[]'
            ptr = c_name + '_click'
            extra_configs.append({
                'struct_type': 'uint8_t',
                'struct_var': name.upper(),
                'struct_init': multi_click_counts,
            })
            events_cfg['multi_click'] = {
                'click_counts': ptr.upper(),
                'count': len(multi_click_counts)
            }
        else:
            raise ValueError('YAML validation error in  button device: Multi-click value must be an array.')

    if enabled_events['long_press_start'] == 1:
        # Handle long_press_start_time array
        long_press_start_times = events_config.get('long_press_start_time')
        if long_press_start_times is not None and isinstance(long_press_start_times, list):
            long_press_start_durations = []
            long_press_start_durations = [int(x) for x in long_press_start_times]
            name = c_name + '_lp_start[]'
            ptr = c_name + '_lp_start'
            extra_configs.append({
                'struct_type': 'uint16_t',
                'struct_var': name.upper(),
                'struct_init': long_press_start_durations,
            })
            events_cfg['long_press_start'] = {
                'durations_ms': ptr.upper(),
                'count': len(long_press_start_durations)
            }
        elif long_press_start_times is not None and not isinstance(long_press_start_times, list):
            raise ValueError('YAML validation error in  button device: Long press start time value must be an array.')

    if enabled_events['long_press_up'] == 1:
        # Handle long_press_up_time array
        long_press_up_times = events_config.get('long_press_up_time')
        if long_press_up_times is not None and isinstance(long_press_up_times, list):
            long_press_up_durations = []
            long_press_up_durations = [int(x) for x in long_press_up_times]
            name = c_name + '_lp_up[]'
            ptr = c_name + '_lp_up'
            extra_configs.append({
                'struct_type': 'uint16_t',
                'struct_var': name.upper(),
                'struct_init': long_press_up_durations,
            })
            events_cfg['long_press_up'] = {
                'durations_ms': ptr.upper(),
                'count': len(long_press_up_durations)
            }
        elif long_press_up_times is not None and not isinstance(long_press_up_times, list):
            raise ValueError('YAML validation error in  button device: Long press up time value must be an array.')

    # Initialize button configuration structure with nested events structure
    button_config = {
        'sub_type': button_type,
        'button_timing_cfg': button_timing_cfg,
        'events_cfg': events_cfg,
        'sub_cfg': {}
    }

    # Parse GPIO button configuration
    if button_type == 'gpio':
        if 'gpio_name' in device_config:
            raise ValueError(
                "YAML validation error in  button device: Legacy 'config.gpio_name' is no longer supported. "
                "Declare the GPIO peripheral under top-level 'peripherals'."
            )

        gpio_name = _find_named_peripheral(peripherals, 'gpio')
        if gpio_name is None:
            raise ValueError(
                "YAML validation error in  button device: Missing GPIO peripheral in top-level 'peripherals'"
            )

        # Validate GPIO peripheral exists if peripherals_dict is provided
        if peripherals_dict is not None and gpio_name not in peripherals_dict:
            raise ValueError(f" button device {name} references undefined peripheral '{gpio_name}'")

        # Get active level
        active_level = int(device_config.get('active_level', 0))
        if active_level not in [0, 1]:
            raise ValueError(f"YAML validation error in  button device: Invalid active_level '{active_level}'. Valid values: [0, 1]")

        # Get power save configuration
        enable_power_save = bool(device_config.get('enable_power_save', False))

        # Get pull configuration
        disable_pull = bool(device_config.get('disable_pull', False))

        # Build GPIO button configuration
        gpio_cfg = {
            'gpio_name': gpio_name,
            'active_level': active_level,
            'enable_power_save': enable_power_save,
            'disable_pull': disable_pull
        }
        button_config['sub_cfg']['gpio'] = gpio_cfg

    # Parse ADC button configuration
    elif button_type == 'adc_multi' or button_type == 'adc_single':
        if 'adc_name' in device_config:
            raise ValueError(
                "YAML validation error in  button device: Legacy 'config.adc_name' is no longer supported. "
                "Declare the ADC peripheral under top-level 'peripherals'."
            )

        adc_name = _find_named_peripheral(peripherals, 'adc')
        if adc_name is None:
            raise ValueError(
                "YAML validation error in  button device: Missing ADC peripheral in top-level 'peripherals'"
            )

        # Validate ADC peripheral exists if peripherals_dict is provided
        if peripherals_dict is not None and adc_name not in peripherals_dict:
            raise ValueError(f" button device {name} references undefined peripheral '{adc_name}'")

        # Determine configuration based on sub_type
        if button_type == 'adc_multi':
            # Multi-button configuration
            button_num = device_config.get('button_num')
            voltage_range = device_config.get('voltage_range')
            button_labels = device_config.get('button_labels')

            if button_num is None:
                raise ValueError(f'YAML validation error in  button device: Missing button_num field for ADC multi-button configuration')
            if voltage_range is None:
                raise ValueError(f'YAML validation error in  button device: Missing voltage_range field for ADC multi-button configuration')

            button_num = int(button_num)
            if button_num <= 1:
                raise ValueError(f'YAML validation error in  button device: button_num must be > 1 for multi-button configuration')
            if not isinstance(voltage_range, list):
                raise ValueError(f'YAML validation error in  button device: voltage_range must be a list')
            if len(voltage_range) != button_num:
                raise ValueError(f'YAML validation error in  button device: voltage_range length ({len(voltage_range)}) must match button_num ({button_num})')

            max_voltage = device_config.get('max_voltage', 3000)
            voltage_range = [int(v) for v in voltage_range]
            # Validate button_labels if provided
            if button_labels is not None:
                if not isinstance(button_labels, list):
                    raise ValueError(f'YAML validation error in  button device: button_labels must be a list')
                if len(button_labels) != button_num:
                    raise ValueError(f'YAML validation error in  button device: button_labels length ({len(button_labels)}) must match button_num ({button_num})')
                # Keep as list of strings
                button_labels = [f'"{label}"' for label in button_labels]
            else:
                button_labels = []

            # Build ADC button configuration with multi union
            adc_cfg = {
                'adc_name': adc_name,
                'multi': {
                    'button_num': button_num,
                    'voltage_range': voltage_range,
                    'button_labels': button_labels,
                    'max_voltage': max_voltage
                }
            }
        else:  # button_type == 'adc_single'
            # Single-button configuration
            button_index = int(device_config.get('button_index', 0))
            min_voltage = int(device_config.get('min_voltage', 0))
            max_voltage = int(device_config.get('max_voltage', 500))
            if min_voltage >= max_voltage:
                raise ValueError(f'YAML validation error in  button device: min_voltage ({min_voltage}) must be less than max_voltage ({max_voltage})')

            # Build ADC button configuration with single union
            adc_cfg = {
                'adc_name': adc_name,
                'single': {
                    'button_index': button_index,
                    'min_voltage': min_voltage,
                    'max_voltage': max_voltage
                }
            }
        button_config['sub_cfg']['adc'] = adc_cfg

    # Parse Custom button configuration
    elif button_type == 'custom':
        # Custom button does not require specific sub_cfg parameters
        # The device name itself is used to look up the driver
        pass

    # Create the result structure matching ESP Board Manager format
    result = {
        'struct_type': 'dev_button_config_t',
        'struct_var': f'{c_name}_cfg',
        'struct_init': button_config,
        'extra_configs': extra_configs
    }

    return result
